src/test/java/com/exasol/spark/s3/S3CleanupIT.java

```diff
@@ -0,0 +1,226 @@
+        };
+        final Dataset<Integer> df = spark.read() //
+                .format("exasol-s3") //
+                .option("table", table.getFullyQualifiedName()) //
+                .options(getSparkOptions()) //
+                .load() //
+                .map(failAlways, Encoders.INT());
+        final SparkException exception = assertThrows(SparkException.class, () -> df.collectAsList());
+        assertThat(exception.getMessage(), containsString("Intentional failure for all tasks."));
+        assertThatBucketIsEmpty();
+    }
+
+    private Dataset<Row> getSampleSparkDataset() {
+        final StructType schema = new StructType() //
+                .add("c_str", DataTypes.StringType, false) //
+                .add("c_int", DataTypes.IntegerType, false) //
+                .add("c_double", DataTypes.DoubleType, false) //
+                .add("c_bool", DataTypes.BooleanType, false);
+        return spark.createDataFrame(Arrays.asList( //
+                RowFactory.create("str", 10, 3.14, true), //
+                RowFactory.create("abc", 20, 2.72, false) //
+        ), schema);
+    }
+
+    @Test
+    void testSinkSuccessJobEndCleanup() throws SQLException {
+        final Table table = exasolSchema.createTableBuilder("table_cleanup_save") //
+                .column("c_str", "VARCHAR(3)") //
+                .column("c_int", "DECIMAL(9,0)") //
+                .column("c_double", "DOUBLE") //
+                .column("c_bool", "BOOLEAN") //
+                .build();
+        getSampleSparkDataset() //
+                .write() //
+                .mode("append") //
+                .format("exasol-s3") //
+                .options(getSparkOptions()) //
+                .option("table", table.getFullyQualifiedName()) //
+                .save();
+        final String query = "SELECT * FROM " + table.getFullyQualifiedName() + " ORDER BY \"c_int\" ASC";
+        try (final ResultSet result = connection.createStatement().executeQuery(query)) {
+            assertThat(result, table().row("str", 10, 3.14, true).row("abc", 20, 2.72, false).matches());
+        }
+        assertThatBucketIsEmpty();
+    }
+
+    @Test
+    void testSinkJobAlwaysFailsJobEndCleanup() {
+        final DataFrameWriter<Row> df = getSampleSparkDataset() //
+                .write() //
+                .mode("append") //
+                .format("exasol-s3") //
+                .options(getSparkOptions()) //
+                .option("table", "non_existent_table");
+        final Exception exception = assertThrows(ExasolConnectionException.class, () -> df.save());
+        assertThat(exception.getMessage(), containsString("Failure running the import"));
+        assertThatBucketIsEmpty();
+    }
+
+    /**
+     * This class simulates a failure based on the specified {@link #condition} and for maximum number of times
+     * according to value {@link #maxFailures}.
+     */
+    static class FailMax {
```

Comment:
My guess is the tests only run on a single node, otherwise this class might not work not as expected.

Reply:
Yes, correct. It runs on single machine, but with multiple threads (handles each dataframe partitions). We use `local[*]` as Spark master in the integration tests.

Reply:
good, then this works, in distributed mode it wouldn't work, maybe add a comment, that this class needs local mode

Reply:
Done